src/ILCompiler.Compiler/src/Compiler/CompilerTypeSystemContext.TypeInit.cs

```diff
@@ -26,7 +33,7 @@ partial class CompilerTypeSystemContext
         /// </summary>
         public bool HasLazyStaticConstructor(TypeDesc type)
         {
-            return type.HasStaticConstructor && !HasEagerConstructorAttribute(type);
+            return type.HasStaticConstructor && !HasEagerConstructorAttribute(type) && !HasExplicitClassConstructor(type);
```

Comment:
Maybe we should not have these fine grained eager constructors at all, and just have explicit Initialize methods.